Checkout commit: additional user interface (#4120)

EduardoCopat opened the pull request:
Implements #4120

When checking out to a commit, changes the branch name to the short commit sha:

![image](https://user-images.githubusercontent.com/2248490/102853857-07142800-4486-11eb-9889-b04395a07064.png)

It is showing a grey box around it. I didn't do anything in particular to change it, but it looks good 🤷‍♂️?

When you try to stage in this scenario, the following pop-up will show:

![image](https://user-images.githubusercontent.com/2248490/102853415-29597600-4485-11eb-8092-aea8cb8d3631.png)

This will create a new branch and change to it. Feel free to make changes to the wording.

Also tested:

- Regular stage - OK
- Going to repo again - OK, shows that it is currently in a commit
- Switching branch - OK
- Creating branch - OK 
- Deleting branch - OK 
- Cancelling the pop-ups - OK

EduardoCopat commented on src/ui/zcl_abapgit_gui_router.clas.abap:
```diff
@@ -141,7 +141,7 @@ ENDCLASS.
 
 
-CLASS zcl_abapgit_gui_router IMPLEMENTATION.
+CLASS ZCL_ABAPGIT_GUI_ROUTER IMPLEMENTATION.
```
Not sure why this is happening. It is lowercase in my branch: https://github.com/EduardoCopat/abapGit/blob/4120-dev/src/ui/zcl_abapgit_gui_router.clas.abap

mbtools replied:
One of many issues https://github.com/abapGit/abapGit/issues/3741. Just ignore this one.

mbtools reviewed (changes requested):
On the surface, this looks great. However, you saved the commit in the branch field of the persistence. This can't be good and will have unintended side effects (if not now, then sometime later). 

For example, anything calling `get_selected_branch` is expecting a branch and not a commit:
https://github.com/abapGit/abapGit/blob/master/src/zcl_abapgit_repo_online.clas.abap#L187 

The proper place to change the display is `render_branch_name`:
https://github.com/abapGit/abapGit/blob/master/src/ui/zcl_abapgit_gui_chunk_lib.clas.abap#L273

mbtools commented:
We have a similar check when a tag is checked out. There we simply raise an exception:
https://github.com/abapGit/abapGit/blob/master/src/zcl_abapgit_repo_online.clas.abap#L215
This is where I would put the logic for blocking push on commits.

EduardoCopat commented:
@mbtools Thank you very much for the review and pointers. I was worried about the side effects.

I moved the logic to the `render_branch_name` method. Also, I removed the pop-up and added a validation on push as suggested:

![image](https://user-images.githubusercontent.com/2248490/103035950-629cfd80-45cd-11eb-891d-129af0e1f009.png)

mbtools commented:
- Render look good. 
- The new exception is good too. It comes late (on commit page) but will prevent the bad call to git for example in case on background or API calls without UI. 
- Sorry if you misunderstood. The popup was fine and helpful since it stops going to the staging page. Let's put it back. Maybe also raise the popup for tags `branch_name CP zif_abapgit_definitions=>c_git_branch-tags`

EduardoCopat commented:
No problems, it is easy to add back. I also added the same validation for tags.

mbtools reviewed (approved):
Excellent. Thanks 👍

larshp commented:
thanks 👍
